In [ ]:
#Task 3 → Data Lifecycle & Lineage

In [4]:
import pandas as pd
import numpy as np
import matplotlib as mtp

In [6]:
df=pd.read_csv("/content/retail_sales_raw.csv")

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125 entries, 0 to 124
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   order_id        125 non-null    object 
 1   order_date      125 non-null    object 
 2   customer_id     125 non-null    object 
 3   customer_name   125 non-null    object 
 4   city            122 non-null    object 
 5   category        125 non-null    object 
 6   product         125 non-null    object 
 7   quantity        123 non-null    float64
 8   unit_price      118 non-null    float64
 9   discount_pct    125 non-null    int64  
 10  payment_method  125 non-null    object 
 11  sales_channel   125 non-null    object 
 12  status          125 non-null    object 
dtypes: float64(2), int64(1), object(10)
memory usage: 12.8+ KB


In [9]:
df.describe()

,quantity,unit_price,discount_pct
count,123.000000,118.000000,125.000000
mean,3.463415,845.508475,9.360000
std,1.569654,440.556364,6.779238
min,1.000000,160.000000,0.000000
25%,2.000000,450.000000,5.000000
50%,3.000000,875.000000,10.000000
75%,5.000000,1150.000000,15.000000
max,6.000000,1700.000000,20.000000


In [12]:
df.columns

Index(['order_id', 'order_date', 'customer_id', 'customer_name', 'city',
       'category', 'product', 'quantity', 'unit_price', 'discount_pct',
       'payment_method', 'sales_channel', 'status'],
      dtype='object')

In [16]:
df.head()

,order_id,order_date,customer_id,customer_name,city,category,product,quantity,unit_price,discount_pct,payment_method,sales_channel,status
0,O0001,2026-02-05,C011,Vikas Tiwari,Ujjain,ELECTRONICS,Wireless Mouse,6.0,NaN,0,cash,online,completed
1,O0002,04-25-2026,C004,Neha Singh,Jabalpur,Electronics,Headphones,1.0,1400.0,15,cod,offline,completed
2,O0003,06-04-2026,C006,Sneha Gupta,Bhopal,electronics,Wireless Mouse,1.0,700.0,20,upi,Offline,Pending
3,O0004,02-28-2026,C005,rohit patel,Ujjain,Grocery,Cooking Oil,1.0,160.0,0,upi,Online,cancelled
4,O0005,12/06/2026,C003,AMIT KHAN,Rewa,Grocery,Rice Bag,5.0,950.0,5,cash,Online,Cancelled


In [20]:
df.duplicated().sum()

np.int64(5)

In [21]:
df[df.duplicated()]

,order_id,order_date,customer_id,customer_name,city,category,product,quantity,unit_price,discount_pct,payment_method,sales_channel,status
120,O0045,30-Mar-2026,C012,Meena Joshi,Bhopal,GROCERY,Cooking Oil,6.0,160.0,0,upi,ONLINE,Completed
121,O0048,13/05/2026,C005,rohit patel,Ujjain,GROCERY,Rice Bag,1.0,950.0,15,cash,ONLINE,Completed
122,O0005,12/06/2026,C003,AMIT KHAN,Rewa,Grocery,Rice Bag,5.0,950.0,5,cash,Online,Cancelled
123,O0056,2026-03-16,C002,Priya Verma,Indore,home,Water Bottle,2.0,450.0,10,COD,ONLINE,pending
124,O0027,03-05-2026,C007,Arjun Mehta,Indore,ELECTRONICS,Wireless Mouse,2.0,600.0,15,Card,offline,cancelled


In [24]:
df_clean = df.drop_duplicates().copy()

In [29]:
df_clean.duplicated().sum()

np.int64(0)

In [38]:
print("raw data rows:",len(df))
print("clean data rows:",len(df_clean))

raw data rows: 125
clean data rows: 120


In [40]:
df.duplicated().sum()

np.int64(5)

In [41]:
df_clean.duplicated().sum()

np.int64(0)

In [42]:
df.shape

(125, 13)

In [43]:
df_clean.shape

(120, 13)

In [44]:
df.isnull().sum()

,0
order_id,0
order_date,0
customer_id,0
customer_name,0
city,3
category,0
product,0
quantity,2
unit_price,7
discount_pct,0


In [45]:
df_clean.isnull().sum()

,0
order_id,0
order_date,0
customer_id,0
customer_name,0
city,3
category,0
product,0
quantity,2
unit_price,7
discount_pct,0


In [46]:
print("Raw Data Shape:", df.shape)
print("Clean Data Shape:", df_clean.shape)

print("\nRaw Missing Values:")
print(df.isnull().sum())

print("\nClean Missing Values:")
print(df_clean.isnull().sum())

Raw Data Shape: (125, 13)
Clean Data Shape: (120, 13)

Raw Missing Values:
order_id          0
order_date        0
customer_id       0
customer_name     0
city              3
category          0
product           0
quantity          2
unit_price        7
discount_pct      0
payment_method    0
sales_channel     0
status            0
dtype: int64

Clean Missing Values:
order_id          0
order_date        0
customer_id       0
customer_name     0
city              3
category          0
product           0
quantity          2
unit_price        7
discount_pct      0
payment_method    0
sales_channel     0
status            0
dtype: int64


In [47]:
df[df['city'].isnull()]

,order_id,order_date,customer_id,customer_name,city,category,product,quantity,unit_price,discount_pct,payment_method,sales_channel,status
50,O0051,05-31-2026,C003,AMIT KHAN,NaN,HOME,Water Bottle,1.0,450.0,10,COD,Offline,cancelled
77,O0078,2026-02-26,C002,Priya Verma,NaN,Home,Bedsheet,6.0,1250.0,10,cod,Offline,Cancelled
93,O0094,23-Jun-2026,C009,Karan Jain,NaN,electronics,Keyboard,5.0,900.0,0,upi,Offline,cancelled


In [48]:
df[df['quantity'].isnull()]

,order_id,order_date,customer_id,customer_name,city,category,product,quantity,unit_price,discount_pct,payment_method,sales_channel,status
18,O0019,2026-02-10,C001,Rahul Sharma,Bhopal,Grocery,Tea Pack,NaN,NaN,5,COD,ONLINE,Cancelled
58,O0059,2026-06-22,C012,Meena Joshi,Bhopal,grocery,Cooking Oil,NaN,160.0,20,UPI,Online,completed


In [49]:
df[df['unit_price'].isnull()]

,order_id,order_date,customer_id,customer_name,city,category,product,quantity,unit_price,discount_pct,payment_method,sales_channel,status
0,O0001,2026-02-05,C011,Vikas Tiwari,Ujjain,ELECTRONICS,Wireless Mouse,6.0,NaN,0,cash,online,completed
8,O0009,06-14-2026,C008,Pooja Yadav,Rewa,Grocery,Cooking Oil,1.0,NaN,15,upi,ONLINE,Pending
17,O0018,19-Apr-2026,C003,AMIT KHAN,Rewa,clothing,T-Shirt,2.0,NaN,10,cash,offline,Pending
18,O0019,2026-02-10,C001,Rahul Sharma,Bhopal,Grocery,Tea Pack,NaN,NaN,5,COD,ONLINE,Cancelled
29,O0030,2026-04-22,C009,Karan Jain,Satna,ELECTRONICS,Keyboard,5.0,NaN,5,Cash,Offline,Completed
52,O0053,17-Jan-2026,C008,Pooja Yadav,Rewa,Grocery,Tea Pack,4.0,NaN,20,cash,offline,Completed
111,O0112,01-10-2026,C006,Sneha Gupta,Bhopal,ELECTRONICS,Headphones,4.0,NaN,15,Card,online,completed


In [50]:
df_clean['city'] = df_clean['city'].fillna(df_clean['city'].mode()[0])

In [51]:
df_clean['quantity'] = df_clean['quantity'].fillna(df_clean['quantity'].median())

In [52]:
df_clean['unit_price'] = df_clean['unit_price'].fillna(df_clean['unit_price'].median())

In [53]:
df_clean.isnull().sum()

,0
order_id,0
order_date,0
customer_id,0
customer_name,0
city,0
category,0
product,0
quantity,0
unit_price,0
discount_pct,0


RAW CSV
   │
   ▼
Load into Pandas
   │
   ▼
Data Validation
   │
   ├── 5 Duplicate Rows Found
   │
   ├── City → 3 Missing Values
   │
   ├── Quantity → 2 Missing Values
   │
   └── Unit Price → 7 Missing Values
   │
   ▼
DATA CLEANING / TRANSFORMATION
   │
   ├── Remove Duplicates
   ├── Fill Missing City
   ├── Fill Missing Quantity
   └── Fill Missing Unit Price
   │
   ▼
CLEAN DATA
   │
   ├── Duplicates = 0
   └── Missing Values = 0
   │
   ▼
ANALYSIS
   │
   ▼
KEY METRIC
   │
   ▼
REPORT / DECISION

In [54]:
df_clean[['quantity', 'unit_price', 'discount_pct']].head(10)

,quantity,unit_price,discount_pct
0,6.0,900.0,0
1,1.0,1400.0,15
2,1.0,700.0,20
3,1.0,160.0,0
4,5.0,950.0,5
5,3.0,750.0,5
6,6.0,1700.0,20
7,2.0,280.0,15
8,1.0,900.0,15
9,2.0,160.0,5


In [55]:
df_clean['gross_sales'] = (
    df_clean['quantity'] * df_clean['unit_price']
)

In [56]:
df_clean['discount_amount'] = (
    df_clean['gross_sales'] * df_clean['discount_pct'] / 100
)

In [57]:
df_clean['net_sales'] = (
    df_clean['gross_sales'] - df_clean['discount_amount']
)

In [58]:
df_clean[['quantity',
          'unit_price',
          'discount_pct',
          'gross_sales',
          'discount_amount',
          'net_sales']].head(10)

,quantity,unit_price,discount_pct,gross_sales,discount_amount,net_sales
0,6.0,900.0,0,5400.0,0.0,5400.0
1,1.0,1400.0,15,1400.0,210.0,1190.0
2,1.0,700.0,20,700.0,140.0,560.0
3,1.0,160.0,0,160.0,0.0,160.0
4,5.0,950.0,5,4750.0,237.5,4512.5
5,3.0,750.0,5,2250.0,112.5,2137.5
6,6.0,1700.0,20,10200.0,2040.0,8160.0
7,2.0,280.0,15,560.0,84.0,476.0
8,1.0,900.0,15,900.0,135.0,765.0
9,2.0,160.0,5,320.0,16.0,304.0


Raw CSV
   ↓
Data Validation
   ↓
Duplicate Removal
   ↓
Missing Value Handling
   ↓
Clean Dataset
   ↓
Quantity + Unit Price
   ↓
Gross Sales
   ↓
Discount %
   ↓
Net Sales
   ↓
Final Sales Metric

In [59]:
total_net_sales = df_clean['net_sales'].sum()

print("Total Net Sales:", total_net_sales)

Total Net Sales: 323569.5


In [60]:
print("Sum of individual Net Sales:", df_clean['net_sales'].sum())
print("Total Net Sales:", total_net_sales)

print("Validation:",
      df_clean['net_sales'].sum() == total_net_sales)

Sum of individual Net Sales: 323569.5
Total Net Sales: 323569.5
Validation: True


1. RAW DATA SOURCE
   retail_sales
          ↓
2. DATA INGESTION
   CSV → Pandas DataFrame
          ↓
3. DATA VALIDATION
   Duplicates + Missing Values + Data Types
          ↓
4. DATA CLEANING
   Remove 5 duplicates
   Fill missing values
          ↓
5. DATA TRANSFORMATION
   Gross Sales
   Discount Amount
   Net Sales
          ↓
6. ANALYSIS / REPORTING
   Total Net Sales
   ₹323,569.50
          ↓
7. BUSINESS DECISION
   Sales performance analysis
   → Reporting / Decision making

# Task 3 — Source of Truth & Data Lineage

## 1. Project Overview

This document describes the data lifecycle and lineage of the sales dataset used in the PlaceMux Data Analyst Phase 1 project.

The lifecycle was traced from the original CSV source through data validation, cleaning, transformation, analysis, and final reporting.

## 2. Data Source

The original source is the Task 2 raw sales CSV file.

The dataset contains the following key fields:

* order_id
* order_date
* customer_id
* customer_name
* city
* category
* product
* quantity
* unit_price
* discount_pct
* payment_method
* sales_channel
* status

The raw dataset contained 125 rows and 13 columns.

## 3. Data Validation

The raw dataset was checked for duplicate records and missing values.

Findings:

* 5 duplicate rows were identified.
* Missing values were found in:

  * city: 3
  * quantity: 2
  * unit_price: 7

These issues were addressed during the cleaning stage.

## 4. Data Cleaning

A cleaned dataset named `df_clean` was created.

Cleaning actions included:

* Removing 5 duplicate rows.
* Filling missing city values using the mode.
* Filling missing quantity values using the median.
* Filling missing unit_price values using the median.

After cleaning, the dataset contained:

* No duplicate rows.
* No missing values.

Therefore, `df_clean` was used as the analytical source of truth for this project.

## 5. Data Transformation

The key business metric selected for lineage validation was Net Sales.

The calculation was performed in three stages.

### Gross Sales

`Gross Sales = Quantity × Unit Price`

### Discount Amount

`Discount Amount = Gross Sales × Discount % / 100`

### Net Sales

`Net Sales = Gross Sales − Discount Amount`

## 6. Source of Truth

For this project, `df_clean` is the source of truth for analytical calculations because it is the cleaned and validated version of the original dataset.

The original CSV remains the source of the raw records, while `df_clean` is the source used for analysis and reporting.

## 7. Key Metric Validation

The final calculated metric was:

**Total Net Sales = ₹323,569.50**

The metric was independently validated by comparing the calculated total with the sum of all individual `net_sales` values.

Validation result:

**True**

This confirms that the final metric can be traced from the underlying transaction-level data through the documented transformation steps.

## 8. Data Ownership

| Stage                  | Owner                      |
| ---------------------- | -------------------------- |
| Raw data source        | Data source / sales system |
| Data loading           | Data Analyst               |
| Data validation        | Data Analyst               |
| Data cleaning          | Data Analyst               |
| Data transformation    | Data Analyst               |
| Analysis and reporting | Data Analyst               |
| Business decision      | Business / Management      |

## 9. Retention and Privacy

Source and analytical data should be retained according to the applicable project or company data-retention policy.

Customer-related fields should be protected and accessed only when required for the analysis.

When the applicable retention period ends, data should be securely archived or removed according to the relevant policy.

## 10. End-to-End Lineage

The complete lineage is:

**Raw CSV → Pandas DataFrame → Validation → Cleaning → `df_clean` → Gross Sales → Discount Amount → Net Sales → Reporting → Business Decision**

This demonstrates how a business metric can be traced from its original data source to its final use.
